File: examples/go/03_vector_search.go
// Vector Search Example
//
// Demonstrates vector similarity search:
// - Creating a vector index with HNSW
// - Bulk loading embeddings
// - Finding nearest neighbors
// - Distance utilities

package main

import (
	"fmt"
	"log"

	toondb "github.com/toondb/toondb-go"
)

func main() {
	// Configuration for the vector index
	config := toondb.VectorIndexConfig{
		Dimension:      128,                      // Embedding dimension
		Metric:         toondb.DistanceMetricCosine, // Cosine similarity
		M:              16,                       // HNSW connections per node
		EfConstruction: 100,                      // Construction quality factor
	}

	index, err := toondb.NewVectorIndex("./vector_index", config)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Println("✓ Vector index created")

	// Generate sample embeddings (in practice, use a real embedding model)
	vectors := make([][]float32, 100)
	labels := make([]string, 100)

	for i := 0; i < 100; i++ {
		// Create a simple pattern-based embedding for demo
		vec := make([]float32, 128)
		for j := 0; j < 128; j++ {
			vec[j] = float32((i*j)%256) / 255.0
		}
		vectors[i] = vec
		labels[i] = fmt.Sprintf("document_%d", i)
	}

	// Bulk build the index
	err = index.BulkBuild(vectors, labels)
	if err != nil {
		log.Fatal(err)
	}
	fmt.Printf("✓ Indexed %d vectors\n", len(vectors))

	// Create a query vector (similar to document_42)
	query := make([]float32, 128)
	for j := 0; j < 128; j++ {
		query[j] = float32((42*j)%256) / 255.0
	}

	// Search for nearest neighbors
	results, err := index.Query(query, 5) // k=5
	if err != nil {
		log.Fatal(err)
	}

	fmt.Println("\n✓ Top 5 nearest neighbors:")
	for i, result := range results {
		label := "<no label>"
		if result.Label != "" {
			label = result.Label
		}
		fmt.Printf("  %d. %s (distance: %.4f)\n", i+1, label, result.Distance)
	}

	// Distance utility functions
	a := []float32{1.0, 0.0, 0.0}
	b := []float32{0.707, 0.707, 0.0}

	cosineDist := toondb.ComputeCosineDistance(a, b)
	euclideanDist := toondb.ComputeEuclideanDistance(a, b)

	fmt.Println("\n✓ Distance calculations:")
	fmt.Printf("  Cosine distance: %.4f\n", cosineDist)
	fmt.Printf("  Euclidean distance: %.4f\n", euclideanDist)

	// Normalize a vector
	v := []float32{3.0, 4.0}
	normalized := toondb.NormalizeVector(v)
	fmt.Printf("  Normalized [3, 4]: [%.2f, %.2f]\n", normalized[0], normalized[1])
}

File: examples/rust/03_vector_search.rs
//! Vector Search Example
//! 
//! This example demonstrates vector similarity search:
//! - Creating a vector index with HNSW
//! - Bulk loading embeddings
//! - Finding nearest neighbors
//! - Distance metrics

use sochdb::{VectorIndex, VectorIndexConfig, DistanceMetric};
use anyhow::Result;

fn main() -> Result<()> {
    // Configuration for the vector index
    let config = VectorIndexConfig {
        dimension: 128,                    // Embedding dimension
        metric: DistanceMetric::Cosine,    // Cosine similarity
        m: 16,                             // HNSW connections per node
        ef_construction: 100,              // Construction quality factor
        ef_search: 50,                     // Search quality factor
    };

    let index = VectorIndex::new("./vector_index", config)?;
    println!("✓ Vector index created");

    // Generate sample embeddings (in practice, use a real embedding model)
    let mut vectors: Vec<Vec<f32>> = Vec::new();
    let mut labels: Vec<String> = Vec::new();
    
    for i in 0..100 {
        // Create a simple pattern-based embedding for demo
        let mut vec = vec![0.0f32; 128];
        for j in 0..128 {
            vec[j] = ((i * j) % 256) as f32 / 255.0;
        }
        vectors.push(vec);
        labels.push(format!("document_{}", i));
    }

    // Bulk build the index
    index.bulk_build(&vectors, Some(&labels))?;
    println!("✓ Indexed {} vectors", vectors.len());

    // Create a query vector
    let mut query = vec![0.0f32; 128];
    for j in 0..128 {
        query[j] = ((42 * j) % 256) as f32 / 255.0;  // Similar to document_42
    }

    // Search for nearest neighbors
    let results = index.query(&query, 5)?;  // k=5

    println!("\n✓ Top 5 nearest neighbors:");
    for (i, result) in results.iter().enumerate() {
        println!("  {}. {} (distance: {:.4})", 
            i + 1,
            result.label.as_deref().unwrap_or("<no label>"),
            result.distance);
    }

    // Distance utility functions
    let a = vec![1.0, 0.0, 0.0];
    let b = vec![0.707, 0.707, 0.0];

    let cosine_dist = sochdb::vector::cosine_distance(&a, &b);
    let euclidean_dist = sochdb::vector::euclidean_distance(&a, &b);

    println!("\n✓ Distance calculations:");
    println!("  Cosine distance: {:.4}", cosine_dist);
    println!("  Euclidean distance: {:.4}", euclidean_dist);

    // Normalize a vector
    let v = vec![3.0, 4.0];
    let normalized = sochdb::vector::normalize(&v);
    println!("  Normalized [3, 4]: {:?}", normalized);

    Ok(())
}
